docs/proposals/volume-provisioning.md

```diff
@@ -126,30 +129,247 @@ We propose that:
+       storage asset there, as appropriate StorageClass instance may not exist
+       when the volume will be deleted.
+
+     * `pv.Labels` MUST be set to match `claim.spec.selector`. The provisioner
+       MAY add additional labels.
+
+       *So the created PV matches the claim.*
+
+     * `pv.Spec` MUST be set to match requirements in `claim.Spec`, especially
+       access mode and PV size. The provisioned volume size MUST NOT be smaller
+       than size requested in the claim, however it MAY be larger.
+
+       *So the created PV matches the claim.*
+
+     * `pv.Spec.PersistentVolumeSource` MUST be set to point to the created
+       storage asset.
+
+     * `pv.Spec.PersistentVolumeReclaimPolicy` SHOULD be set to `Delete` unless
+       user manually configures other reclaim policy.
+
+     * `pv.Name` MUST be unique. Internal provisioners use name based on
+       `claim.UID` to produce conflicts when two provisioners accidentally
+       provision a PV for the same claim, however external provisioners can use
+       any mechanism to generate an unique PV name.
+
+  Example of a claim that is to be provisioned by an external provisioner for
+  `foo.org/foo-volume`:
+
+  ```yaml
+  apiVersion: v1
+  kind: PersistentVolumeClaim
+  metadata:
+    annotations:
+      volume.beta.kubernetes.io/storage-class: myClass
+      volume.beta.kubernetes.io/storage-provisioner: foo.org/foo-volume
+    name: fooclaim
+    namespace: default
+    resourceVersion: "53"
+    uid: 5a294561-7e5b-11e6-a20e-0eb6048532a3
+  spec:
+    accessModes:
+    - ReadWriteOnce
+    resources:
+      requests:
+        storage: 4Gi
+  #  volumeName: must be empty!
+  ```
+
+  Example of the created PV:
+
+  ```yaml
+  apiVersion: v1
+  kind: PersistentVolume
+  metadata:
+    annotations:
+      pv.kubernetes.io/provisioned-by: foo.org/foo-volume
+      volume.beta.kubernetes.io/storage-class: myClass
+      foo.org/provisioner: "any other annotations as needed"
+    labels:
+        foo.org/my-label: "any labels as needed"
+    generateName: "foo-volume-"
+  spec:
+    accessModes:
+    - ReadWriteOnce
+    awsElasticBlockStore:
+      fsType: ext4
+      volumeID: aws://us-east-1d/vol-de401a79
+    capacity:
+      storage: 4Gi
+    claimRef:
+      apiVersion: v1
+      kind: PersistentVolumeClaim
+      name: fooclaim
+      namespace: default
+      resourceVersion: "53"
+      uid: 5a294561-7e5b-11e6-a20e-0eb6048532a3
+    persistentVolumeReclaimPolicy: Delete
+  ```
+
+  As result, Kubernetes has a PV that represents the storage asset and is bound
+  to the claim. When everything went well, Kubernetes completed binding of the
+  claim to the PV.
+
+  Kubernetes was not blocked in any way during the provisioning and could
+  either bound the claim to another PV that was created by user or even the
+  claim may have been deleted by the user. In both cases, Kubernetes will mark
+  the PV to be delete using the protocol below.
```

Comment:
There are a couple of other race conditions here, I think, as hinted at by the paragraph below.  Specifically,

1. The claim can be changed to point to a storage class that uses a different provisioner.
2. The claim can be resized to request more storage than it originally did.
3. The claim can be resized to request less storage than it originally did.

Item 3 is a non-issue (the claim will bind to the PVC), but items 1 and 2 will likely cause problems.  I'm unsure what the desired semantics for item 1 should be, as there'll be nothing to stop the binding from occurring, even though the user no longer wants to use the external provisioner.  In the case of item 2, the binding won't happen, and the incorrectly-sized volume should probably be deleted and the binding retried.